### PR TITLE
Update chat model standard tests

### PR DESCRIPTION
**Pull request by baskaryan**
- Refactor standard test classes to make them easier to configure
- Update openai to support stop_sequences init param
- Update groq to support stop_sequences init param
- Update fireworks to support max_retries init param
- Update ChatModel.bind_tools to type tool_choice
- Update groq to handle tool_choice="any". **this may be controversial**

### CODE REVIEW

File: libs/core/langchain_core/language_models/chat_models.py [hunk 1]
```diff
@@ -946,6 +946,8 @@ def dict(self, **kwargs: Any) -> Dict:
     def bind_tools(
         self,
         tools: Sequence[Union[Dict[str, Any], Type[BaseModel], Callable, BaseTool]],
+        *,
+        tool_choice: Union[Literal["any", "auto", "none"], str, None] = None,
```
Comment (ccurme) [inline libs/core/langchain_core/language_models/chat_models.py:950]: (nit) might make sense to do this in its own PR

File: libs/partners/openai/langchain_openai/chat_models/base.py [hunk 2]
```diff
@@ -345,6 +345,8 @@ class BaseChatOpenAI(BaseChatModel):
     http_async_client: Union[Any, None] = None
     """Optional httpx.AsyncClient. Only used for async invocations. Must specify 
         http_client as well if you'd like a custom client for sync invocations."""
+    stop: Optional[Union[List[str], str]] = Field(default=None, alias="stop_sequences")
```
Comment (ccurme) [inline libs/partners/openai/langchain_openai/chat_models/base.py:348]: If I initialize ChatOpenAI with `stop` now, I get a warning:
```
UserWarning: WARNING! stop is not default parameter.
stop was transferred to model_kwargs.
Please confirm that stop is what you intended.
```
Are we un-doing a change here?

**Review by ccurme — COMMENTED** [review]
changes to standard tests LGTM but would like @efriis's opinion

File: libs/core/langchain_core/language_models/chat_models.py (hunk 1 above)
Comment (baskaryan) [reply libs/core/langchain_core/language_models/chat_models.py:950]: yea i got a bit carried away

File: libs/partners/openai/langchain_openai/chat_models/base.py (hunk 2 above)
Comment (baskaryan) [reply libs/partners/openai/langchain_openai/chat_models/base.py:348]: i think this should have opposite effect?

File: libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py [hunk 3]
```diff
@@ -1,74 +1,31 @@
...
-from typing import Type
 
 import pytest
 from langchain_core.language_models import BaseChatModel
 from langchain_core.messages import AIMessage, AIMessageChunk, HumanMessage, ToolMessage
-from langchain_core.pydantic_v1 import BaseModel, Field
-from langchain_core.tools import tool
 
+from langchain_standard_tests.unit_tests.chat_models import (
+    ChatModelTests,
```
Comment (efriis) [inline libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py:8]: why subclass? Can't we treat them separately?

If anything should be shared, maybe do a base class?

File: libs/standard-tests/langchain_standard_tests/unit_tests/chat_models.py [hunk 4]
```diff
@@ -18,81 +21,109 @@ def my_adder_tool(a: int, b: int) -> int:
...
+
     @pytest.fixture
-    def chat_model_has_tool_calling(
-        self, chat_model_class: Type[BaseChatModel]
+    def model(self) -> BaseChatModel:
+        return self.chat_model_class(
+            **{**self.standard_chat_model_params, **self.chat_model_params}
+        )
+
+    @property
```
Comment (efriis) [inline libs/standard-tests/langchain_standard_tests/unit_tests/chat_models.py:50]: also open to making these abstract, so user _has_ to override them

**Review by efriis — COMMENTED** [review]
I think having some extra duplicate stuff between the stubs for each is probably good / verbosity is good in tests like this

Happy to discuss though. main thing is that we probably shouldn't copy all the unit tests into the integration tests.

File: libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py (hunk 3 above)
Comment (baskaryan) [reply libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py:8]: that is what this is

Comment (baskaryan) [reply libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py:8]: see ChatModelUnitTests for unit tests

Comment (efriis) [reply libs/standard-tests/langchain_standard_tests/integration_tests/chat_models.py:8]: lol facepalm

File: libs/partners/ai21/tests/integration_tests/test_standard.py [hunk 5]
```diff
@@ -10,98 +10,38 @@
...
-            chat_model_class,
-            chat_model_params,
-        )
+    def test_usage_metadata(self, model: BaseChatModel) -> None:
+        super().test_usage_metadata(model)
 
-    @pytest.fixture
+
+class TestAI21J2(BaseTestAI21):
+    @property
```
Comment (efriis) [inline libs/partners/ai21/tests/integration_tests/test_standard.py:36]: other option would be a pytest.mark.parameterize on the two diff models in one class

**Review by efriis — APPROVED** [review]
works for me. Only downside of property vs pytest.fixture is it's harder to parameterize (see comment on AI21). IMO property is and duplicated class is more readable so I like this

This will immediately break tests in other repos so should fix em on the quick end

File: libs/partners/groq/langchain_groq/chat_models.py [hunk 6]
```diff
@@ -461,8 +464,6 @@ def _create_message_dicts(
     ) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
         params = self._default_params
         if stop is not None:
-            if "stop" in params:
-                raise ValueError("`stop` found in both the input and default params.")
```
Comment (ccurme) [inline libs/partners/groq/langchain_groq/chat_models.py:465]: Do we need this for openai as well?
```python
from langchain_openai import ChatOpenAI

model = ChatOpenAI()

model.invoke("hi", stop=["you"])
```
```
ValueError: `stop` found in both the input and default params.
```

File: libs/standard-tests/langchain_standard_tests/unit_tests/chat_models.py [hunk 7]
```diff
@@ -18,81 +21,109 @@ def my_adder_tool(a: int, b: int) -> int:
...
+class ChatModelTests(ABC):
+    @property
     @abstractmethod
-    @pytest.fixture
     def chat_model_class(self) -> Type[BaseChatModel]:
         ...
 
-    @pytest.fixture
+    @property
     def chat_model_params(self) -> dict:
```
Comment (ccurme) [inline libs/standard-tests/langchain_standard_tests/unit_tests/chat_models.py:31]: do these need to be fixtures? I'm getting `fixture 'chat_model_class' not found` on `test_stop_sequence`.
